File: Solution-Go/Binary_102.go
// Iteration
func levelOrder(root *TreeNode) [][]int {
    if root == nil {
        return [][]int{}
    }
    res, q := [][]int{}, []*TreeNode{}
    q = append(q, root)
    for len(q) > 0 {
        layer := []int{}
        for i := len(q); i > 0; i-- {
            node := q[0]
            layer = append(layer, node.Val)
            q = q[1:]
            if node.Left != nil {
                q = append(q, node.Left)
            }
            if node.Right != nil {
                q = append(q, node.Right)
            }
        }
        res = append(res, layer)
    }
    return res
}



// Recursion. Similar with 107.
func levelOrder(root *TreeNode) [][]int {
    if root == nil {
        return [][]int{}
    }
    res := [][]int{}
    levelMaker(&res, root, 0)
    return res
}

func levelMaker(res *[][]int, root *TreeNode, depth int) {
    if depth >= len(*res) {
        *res = append(*res, []int{})
    }
    (*res)[depth] = append((*res)[depth], root.Val)
    if root.Left != nil {
        levelMaker(res, root.Left, depth + 1)
    }
    if root.Right != nil {
        levelMaker(res, root.Right, depth + 1)
    }
}

File: Solution-Python/Binary_102.py
from collections import deque

class Solution:
    # Iteration
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []
        res = []
        q = deque([root])
        while q:
            layer = []
            for i in range(len(q)):
                node = q.popleft()
                layer.append(node.val)
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            res.append(layer)
        return res



    # Recursion. Similar with 107.
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []
        res = []
        self.levelMaker(res, root, 0)
        return res
    
    def levelMaker(self, res: List[List[int]], root: TreeNode, depth: int):
        if depth >= len(res):
            res.append([])
        res[depth].append(root.val)
        if root.left:
            self.levelMaker(res, root.left, depth + 1)
        if root.right:
            self.levelMaker(res, root.right, depth + 1)